### DISCUSSION TITLE
Allow clients to override the Offline Session Max Limited realm level setting

### DISCUSSION THREAD

**Discussion by NZnick** [keycloak/keycloak#28571] [Ideas]
Currently in Keycloak is it possible to have Keycloak issue refresh tokens with an unlimited lifetime to OIDC clients as long as the `Offline Session Max Limited` flag in the realm settings is turned off (in fact this is the default behaviour). However, because this setting is applied at the realm level one must choose between having all refresh tokens with an unlimited lifetime or having all refresh tokens issued with a limited lifetime. In addition, when `Offline Session Max Limited` is enabled, the maximum lifetime of the refresh tokens is limited to the `SSO Session Max` time, which is also set at the realm level (feel free to correct me if I am wrong).

Each OIDC client in the realm serves a different purpose, there are some cases where it makes good business sense for Keycloak to issue refresh tokens with an unlimited lifetime to some clients while issuing refresh tokens with a limited lifetime to other clients. It is possible to have clients override the session time settings that are defined at the realm level so you can allow refresh tokens issued to some clients to have a longer lifetime than tokens issued to other clients but this is not the same as unlimited refresh token lifetimes for some clients and limited lifetimes for others.

Why is this setting defined at a realm level only with no override on the client level like the other session settings found under a client's advanced settings?

I would like to use this discussion to see if it would be possible to allow clients to override the `Offline Session Max Limited` realm level setting so that some clients can be issued with refresh tokens that have an unlimited lifetime while other clients have their token lifetimes limited by the realm setting.

**Comment by djordjemapp** [comment]
Did you tried to change this settings under client properties? Or there are for client service accounts? 
<img width="769" alt="Screenshot 2024-04-12 at 10 14 28" src="https://github.com/keycloak/keycloak/assets/165149359/a9520806-0a71-43cf-bb48-00883b0989e0">

**Reply by NZnick** [reply]
Yes, we have tried changing those settings, That is what I mean when I say 

> It is possible to have clients override the session time settings that are defined at the realm level so you can allow refresh tokens issued to some clients to have a longer lifetime than tokens issued to other clients but this is not the same as unlimited refresh token lifetimes for some clients and limited lifetimes for others.

The problem is that being able to override the Offline Session Idle and Max settings at a client level to have the refresh tokens that are issued to some clients being longer or shorter than the realm default is not the same as being able to instruct Keycloak to issue refresh tokens that do not expire (have an unlimited lifetime) to some clients while limiting the lifetime of the refresh tokens for other clients. 

At the moment, Keycloak either issues refresh tokens with **no limit to all clients** or issues refresh tokens with **some limit to all clients**. The settings you show above only allow you to _change what that limit is_ on a per-client basis. They do not allow you to specify that a particular client's refresh tokens should be issued with _no limit_.

In addition, the limits you set here cannot exceed the realm's `SSO Session Max` time which means that you can't set values here that would make the refresh token's lifetime very long while also having a reasonably short `SSO Session Max` time as you can if the realm level Offline Session Max Limited` toggle is off.

My idea would be to have a toggle in the same Advanced Settings section above that allows you to specify that a particular client should override what the realm level Offline Session Max Limited` is set to (meaning that refresh token lifetimes for a given client should be unlimited even if the realm level setting is to limit the lifetime and vice-versa).
